### PR TITLE
Dynamic shape mem reuse solution

### PR DESCRIPTION
**Pull request by riverlijunjie**
### Details:
- Dynamic shape inference memory reuse solution
In the current implementation, each memory object allocate its own memory on the heap, without memory reuse across all the graph tensors as it is done in case of static shapes. Using separated memory means bad memory locality and extra memory consumption. This problem can be addressed by using some smart memory reuse technique adopted for memory reuse in dynamic shapes case.

### Tickets:
 - 80804

### CODE REVIEW

File: src/plugins/intel_cpu/src/edge.cpp [hunk 1]
```diff
@@ -272,11 +272,30 @@ void Edge::allocate(const void* mem_ptr) {
...
     memoryPtr.reset(new Memory(parentPtr->getEngine()));
-
-    memoryPtr->Create(inputDesc, mem_ptr, false);  // no pads zeroing
+    allocate(memoryPtr, inputDesc);
     status = Status::Allocated;
 }
 
+void Edge::allocate(const void* mem_ptr) {
+    auto allocateFunc = [&](const MemoryPtr& memoryPtr, const MemoryDesc& inputDesc) {
+        memoryPtr->Create(inputDesc, mem_ptr, false);  // no pads zeroing
```
Comment (t-jankowski) [inline src/plugins/intel_cpu/src/edge.cpp:281]: Reference in lambda to local variable `mem_ptr` is risky. It doesn't fail in this case as `allocateCommon` calls the lambda before it's destroyed. Consider 'by value'.

File: src/plugins/intel_cpu/src/edge.cpp [hunk 2]
```diff
@@ -272,11 +272,30 @@ void Edge::allocate(const void* mem_ptr) {
...
+    allocateCommon(allocateFunc);
+}
+
+void Edge::allocate(DnnlMemoryMngrPtr memMngr) {
+    if (!memMngr) {
+        IE_THROW(Unexpected) << "Memory manager ptr is NULL";
+    }
+
+    auto allocateFunc = [&](const MemoryPtr& memoryPtr, const MemoryDesc& inputDesc) {
+        memoryPtr->Create(inputDesc, memMngr);
```
Comment (t-jankowski) [inline src/plugins/intel_cpu/src/edge.cpp:293]: Consider passing `memMngr` by value .. refer my comment few lines above.

File: src/plugins/intel_cpu/src/memory_desc/cpu_memory_desc_utils.cpp [hunk 3]
```diff
@@ -98,6 +98,14 @@ InferenceEngine::Blob::Ptr MemoryDescUtils::interpretAsBlob(const Memory &mem) {
     return make_blob_with_precision(desc, mem.GetData());
 }
 
+InferenceEngine::TensorDesc MemoryDescUtils::interpretAsBlobDesc(const Memory &mem) {
+    auto& memDesc = mem.getDesc();
+    InferenceEngine::TensorDesc desc = convertToTensorDesc(memDesc);
+
+    desc = InferenceEngine::TensorDesc(desc.getPrecision(), memDesc.getShape().getStaticDims(), desc.getBlockingDesc());
+    return desc;
+}
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/plugins/intel_cpu/src/memory_desc/cpu_memory_desc_utils.cpp:107]: ok

File: src/plugins/intel_cpu/src/cpu_memory.h [hunk 4]
```diff
@@ -17,6 +17,7 @@
 #include <functional>
 #include <memory>
 #include <vector>
+#include <mutex>
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/plugins/intel_cpu/src/cpu_memory.h:20]: sure.

File: src/plugins/intel_cpu/src/edge.cpp (hunk 1 above)
Comment (riverlijunjie) [reply src/plugins/intel_cpu/src/edge.cpp:281]: yes, you are right.

File: src/plugins/intel_cpu/src/edge.cpp (hunk 2 above)
Comment (riverlijunjie) [reply src/plugins/intel_cpu/src/edge.cpp:293]: ok

**Review by t-jankowski — APPROVED** [review]

**Review by tomdol — APPROVED** [review]

File: src/plugins/intel_cpu/src/infer_request.cpp [hunk 5]
```diff
@@ -431,12 +431,8 @@ void LegacyInferRequest::SetBlob(const std::string& name, const InferenceEngine:
                 IE_THROW(ParameterMismatch) << "Failed to set input blob. Blocking descriptor mismatch.";
             }
 
-            auto pBlob = MemoryDescUtils::interpretAsBlob(graph->getInputNodeByName(name)->getChildEdgesAtPort(0)[0]->getMemory());
-            if (!pBlob) {
-                IE_THROW() << "Blob returned after trying to interpret input node's memory is nullable. Input node name: " << name;
-            }
-
-            if (data->getTensorDesc() == pBlob->getTensorDesc() &&
+            auto pBlobDesc = MemoryDescUtils::interpretAsBlobDesc(graph->getInputNodeByName(name)->getChildEdgesAtPort(0)[0]->getMemory());
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/plugins/intel_cpu/src/infer_request.cpp:434]: 1. It should be non-empty due to each node should have at least one edge.
2. It is called in legacy api, will not impact 2.0 api.

**Review by wangleis — APPROVED** [review]

**Pull request merged**